initial proposal for site-allowed-contracts

NikitaVr opened the pull request:
# Note
Keeping the below description up to date with latest contents of the proposal, so that it's easy to read

# Website well-known Contracts [Draft]

<aside>
🚧 This document is a work in progress, we are gathering feedback and suggestions to arrive at a complete specification.

</aside>

Inspired by Content Security Policy (CSP) [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

# Problem Definition

Currently there is no standard way to tell what contracts a website is expected to call. This means that a hacked site can request a signature for any contract, it can steal your tokens, cause you to sign a multisig compromising an organization, or otherwise damage the user.

Another problem is that any site can call a smart contract, and this means that any site is part of the attack surface. A hack on aj ecommerce store, a blog, or really any web page can lead to a legitimate site draining a user's wallet.

# Proposed Solution

Websites that interact with smart contracts should publish a list that let's wallets and other interfaces know what contracts the website is expected to call.

Any transaction from a site that requests a contract not on the list can then be treated as suspicious, and the user can be warned.

Also, as standards become more widely adopted, sites should be treated as unexpected to call a smart contract unless they explicitly declare it by providing a list.

# Technical Details

## Format

```jsx
//.well-known/contracts
{
  contracts: [
    {
      chainId: "string",
      contracted: "string",
    },
  ];
}
```

## Cases

### No well-known is present

The site is not expected to call any smart contacts. If it does call a smart contract treat it as suspicious. This is how CORS works where by default all External requests are blocked.

```jsx
//.well-known/contracts
undefined;
```

### wild card

Any smart contract is allowed from any chain

```jsx
//.well-known/contracts
{
  contracts: [
    {
      chainId: "*",
      contracted: "*",
    },
  ];
}
```

### Wild Card for Networks/Chains

In this sample only networks on Ethereum (1) and Polygon (137) are allowed.

```jsx
//.well-known/contracts
{
	contracts:
	[
		{
			"chainId": "1",
			"contracted": "*"
		}
		{
			"chainId": "137",
			"contracted": "*"
		}
	]
}
```

You could also set up TestNet IDs instead in your test environment.

### Specific list

Check if a contract and chain are on the list. Any call not on the list should be blocked or at least warned about.

# Wallet implementation

We recommend warning users. Outright blocking similar to CSP can be done once the standard is more widely accepted.

## Early Stage

The contract list is fully checked

Sites without a contract list are not warned about since the standard is still being adopted.

## End Game

Once the standard has been adopted allowing website developers with contracts to have time to at least add a wildcard on their site.

Any site that asks to sign a transaction for a contract without a contract list present should trigger a warning to the user.

Contract list is fully checked

# Open Questions

## .well-known/contracts page vs Using Headers vs <meta> Tag

Is the best way to share and expose this data through a page server on the domain, something like [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts) , or by using headers on the site itself similar to CSP.

Another alternative is using the meta tag, similar to CSP [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

```html
<meta
  http-equiv="Content-Security-Policy"
  content="default-src 'self'; img-src https://*; child-src 'none';"
/>
```

# References

Content Security Policy (CSP) [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

Yearn Finance Allowlist [https://medium.com/iearn/yearn-allowlist-71757d4e3cf4](https://medium.com/iearn/yearn-allowlist-71757d4e3cf4)

Uniswap Token Lists [https://uniswap.org/blog/token-lists](https://uniswap.org/blog/token-lists)

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+# Technical Details
+
+## Format
```
I was expecting a JSON schema with an explanation of each field here

NikitaVr replied:
done

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+You could also set up TestNet IDs instead in your test environment.
+
+### Specific list
```
This should probably the first or second example and should include some sample JSON.

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+### Wild Card for Networks/Chains
+
+In this sample only networks on Ethereum (1) and Polygon (137) are allowed.
```
Maybe include a link somewhere of where to find these IDs for those curious about it

NikitaVr replied:
good idea, added

umar-ahmed commented on proposals/site-allowed-contracts.md:
````diff
@@ -0,0 +1,137 @@
+}
+```
+
+### Wild Card for Networks/Chains
````
```suggestion
### Wildcard specifier for Networks/Chains
```

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+# Wallet implementation
+
+We recommend warning users. Outright blocking similar to CSP can be done once the standard is more widely accepted.
```
Anything future looking is usually discounted in proposals like these without concrete roadmaps. I'd leave this out or be more specific about what you mean by warning users. What types of contracts might a user sign? Where is the right place to warn? Who should be notified?

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+The contract list is fully checked
+
+Sites without a contract list are not warned about since the standard is still being adopted.
```
Chrome does a good job of warning developers that they are missing some security feature and that it will break their sites in a coming version of the browser. I think it might make sense to warn the dev somehow that their contract is not set up right. Maybe the wallet should log something to the console or warn if using a testnet

umar-ahmed commented on proposals/site-allowed-contracts.md:
````diff
@@ -0,0 +1,137 @@
+}
+```
+
+You could also set up TestNet IDs instead in your test environment.
````
Known testnet IDs could just be hard-coded into the wallets themselves. Similar to how Chrome ships with a list of sites to use HSTS on: https://hstspreload.org/ and for root CAs: https://www.chromium.org/Home/chromium-security/root-ca-policy/

NikitaVr replied:
yeh most wallets already keep their own lists of networks. The main point here was to make sure a staging / test version of a website does not accidentally call contracts on a mainnet. Will see how to reword this better.

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+## End Game
+
+Once the standard has been adopted allowing website developers with contracts to have time to at least add a wildcard on their site.
```
Sentence fragment

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+Any site that asks to sign a transaction for a contract without a contract list present should trigger a warning to the user.
+
+Contract list is fully checked
```
idk what this means. I'd just get rid of it

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+Any site that asks to sign a transaction for a contract without a contract list present should trigger a warning to the user.
+
+Contract list is fully checked
+
```
Consider adding a section called "Alternatives Considered" and document what projects are currently doing

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+## .well-known/contracts page vs Using Headers vs <meta> Tag
+
+Is the best way to share and expose this data through a page server on the domain, something like [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts) , or by using headers on the site itself similar to CSP.
```
```suggestion
Is the best way to expose this data via a [well-known URL](https://www.rfc-editor.org/rfc/rfc8615), ex. [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts), or by using HTTP headers on web server that serves the site (similar to CSP).
```

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+Is the best way to share and expose this data through a page server on the domain, something like [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts) , or by using headers on the site itself similar to CSP.
+
+Another alternative is using the meta tag, similar to CSP [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)
```
The meta tag approach isn't really an alternative. It's more a by-product of the `http-equiv` attribute, which allows a web page to tell a user agent to treat it's content as if it is an HTTP header. You can't logically support the <meta> tag approach without the HTTP header approach since you'll break the semantics

NikitaVr replied:
nice got it

umar-ahmed reviewed (commented):
Overall, I like the idea. I'd love to see some more examples of real-world use-cases though. The CSP docs on MDN provide really good examples that help to explain common usage patterns. This is especially helpful because CSP has so many options and attributes, so it's hard to wrap your head around it.

The other thing I think you should "borrow" from CSP is a way to roll this out in an incremental fashion. In CSP, you can say rename the name of the header and it will operation in a "dry-run"/warning mode instead of an outright block. I think something like that would be best to include here rather than hand-waving over the stuff about "eventually this will be standardized and then we'll block stuff properly". In practice, I think that will never happen, so it's better to provide a plan for adoption up front.

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+Currently there is no standard way to tell what contracts a website is expected to call. This means that a hacked site can request a signature for any contract, it can steal your tokens, cause you to sign a multisig compromising an organization, or otherwise damage the user.
+
+Another problem is that any site can call a smart contract, and this means that any site is part of the attack surface. A hack on an ecommerce store, a blog, or really any web page can lead to a legitimate site draining a user's wallet.
```
A website that normally does not interact with a smart contract, ex. most Web2 sites, can be hacked and start initiating transactions.

umar-ahmed commented on proposals/site-allowed-contracts.md:
````diff
@@ -0,0 +1,137 @@
+
+The site is not expected to call any smart contacts. If it does call a smart contract treat it as suspicious. This is how CORS works where by default all External requests are blocked.
+
+```jsx
+//.well-known/contracts
+undefined;
+```
+
````
```suggestion
```

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,137 @@
+
+### No well-known is present
+
+The site is not expected to call any smart contacts. If it does call a smart contract treat it as suspicious. This is how CORS works where by default all External requests are blocked.
```
link to docs on CORS

NikitaVr commented:
Thanks for the really great points. Updated the spec to make it easier to roll out incrementaly, and added in all your feedback. Will be adding more examples as well

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,193 @@
+🚧 This document is a work in progress, we are gathering feedback and suggestions to arrive at a complete specification.
+
+</aside>
+
```
```suggestion

### Abstract

This proposal defines a new ".well-known" web resource (as defined in [RFC-8614](https://www.rfc-editor.org/rfc/rfc8615) for EVM smart contracts. It provides an interface for website authors to hint to crypto wallets what transactions (if any) to expect from a site during a user agent's web browsing session. In doing so, it helps crypto wallets better warn users about unexpected or potentially malicious transactions.

```

umar-ahmed commented on proposals/site-allowed-contracts.md:
````diff
@@ -0,0 +1,193 @@
+
+Check if a contract and chain are on the list. Any call not on the list should be blocked or at least warned about.
+
+```jsx
````
This should be json. And the keys should have quotes around them

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,193 @@
+
+## Contract Details, Functions, and CallData
+
+We chose to keep the spec to just the networkId and contractId to make it easier to agree on the core concept of the list and make the first version easier to implement and adopt.
```
```suggestion
We chose to limit the scope of the spec to just the `networkId` and `contractId` to make it easier to agree on the core concept of the list and make the first version easier to implement and adopt.
```

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,193 @@
+
+We chose to keep the spec to just the networkId and contractId to make it easier to agree on the core concept of the list and make the first version easier to implement and adopt.
+
+Future versions may have more details on the each contract, such as which function calls are expected, similar to the [Yearn Allowlist](https://medium.com/iearn/yearn-allowlist-71757d4e3cf4).
```
```suggestion
Future revisions may have more details on each contract, such as which function calls are expected, similar to the [Yearn Allowlist](https://medium.com/iearn/yearn-allowlist-71757d4e3cf4).
```

umar-ahmed commented on proposals/site-allowed-contracts.md:
```diff
@@ -0,0 +1,193 @@
+
+# Proposed Solution
+
+Websites that interact with smart contracts should publish a list that lets wallets and other interfaces know what contracts the website is expected to call.
```
```suggestion
Websites that interact with smart contracts should publish a list at `/.well-known/contracts` that lets wallets and other interfaces know what contracts the website is expected to call.
```

NikitaVr commented:
This could also prevent people from signing transactions on the wrong network, will consider adding that to the proposal
